File: roadmap-audit.js
const fs = require('fs');
const githubIssues = require('./github-issues.json');

// Extract issue numbers from ROADMAP.md
const roadmapContent = fs.readFileSync('ROADMAP.md', 'utf-8');
const roadmapIssueMatches = roadmapContent.matchAll(/#(\d+)/g);
const roadmapIssues = new Set();
for (const match of roadmapIssueMatches) {
  roadmapIssues.add(parseInt(match[1]));
}

// Get GitHub issue numbers
const githubIssueNumbers = new Set(githubIssues.map(i => i.number));

// Analysis
const phantomIssues = [...roadmapIssues].filter(n => !githubIssueNumbers.has(n)).sort((a,b) => a-b);
const orphanIssues = githubIssues.filter(i => !roadmapIssues.has(i.number)).sort((a,b) => a.number - b.number);

console.log('═══════════════════════════════════════════════════════');
console.log('🎯 ROADMAP AUDIT - EXECUTIVE SUMMARY');
console.log('═══════════════════════════════════════════════════════\n');
console.log('Audit Date:', new Date().toISOString().split('T')[0]);
console.log('Scope:', githubIssues.length, 'GitHub issues vs ROADMAP.md');

// Issue Count Reconciliation
console.log('\n📊 ISSUE COUNT AUDIT');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

const roadmapHeader = roadmapContent.match(/Total de Issues:\s*(\d+)\s*issues\s*\((\d+)\s*abertas?\s*\+\s*(\d+)\s*fechadas?\)/i);
const roadmapTotal = roadmapHeader ? parseInt(roadmapHeader[1]) : 'NOT FOUND';
const roadmapOpen = roadmapHeader ? parseInt(roadmapHeader[2]) : 'NOT FOUND';
const roadmapClosed = roadmapHeader ? parseInt(roadmapHeader[3]) : 'NOT FOUND';

const githubTotal = githubIssues.length;
const githubClosed = githubIssues.filter(i => i.state === 'CLOSED').length;
const githubOpen = githubIssues.filter(i => i.state === 'OPEN').length;

console.log('ROADMAP.md:       ', roadmapTotal, 'issues');
console.log('GitHub (actual):  ', githubTotal, 'issues');
if (roadmapTotal !== 'NOT FOUND') {
  const drift = githubTotal - roadmapTotal;
  const driftPct = ((Math.abs(drift) / githubTotal) * 100).toFixed(1);
  console.log('Drift:            ', drift >= 0 ? '+' : '', drift, 'issues (' + driftPct + '%)');
  console.log('Status:           ', driftPct < 5 ? '✅ OK' : driftPct < 10 ? '🟡 WARNING' : '🔴 CRITICAL');
}

console.log('\nBREAKDOWN:');
console.log('  Documented in ROADMAP:', roadmapIssues.size, 'unique issue numbers');
console.log('  Exist in GitHub:      ', githubIssueNumbers.size, 'issues');
console.log('  Phantom (doc only):   ', phantomIssues.length, 'issues');
console.log('  Orphan (GitHub only): ', orphanIssues.length, 'issues');

// Phantom issues
if (phantomIssues.length > 0) {
  console.log('\n👻 PHANTOM REFERENCES AUDIT');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('CRITICAL:', phantomIssues.length, 'phantom issues detected!\n');

  console.log('Phantom issues:', phantomIssues.map(n => '#'+n).join(', '));

  // Check for ranges
  const ranges = [];
  let start = phantomIssues[0];
  let end = start;
  for (let i = 1; i < phantomIssues.length; i++) {
    if (phantomIssues[i] === end + 1) {
      end = phantomIssues[i];
    } else {
      if (end - start >= 2) {
        ranges.push(`#${start}-#${end}`);
      }
      start = phantomIssues[i];
      end = start;
    }
  }
  if (end - start >= 2) {
    ranges.push(`#${start}-#${end}`);
  }

  if (ranges.length > 0) {
    console.log('\nLikely phantom ranges in ROADMAP:', ranges.join(', '));
    console.log('├─ Reality: These issues might be PR numbers or typos');
    console.log('├─ Impact: ROADMAP references non-existent issues');
    console.log('└─ Action: Search ROADMAP for these patterns and verify');
  }
}

// Orphan issues
if (orphanIssues.length > 0) {
  console.log('\n🆕 ORPHAN ISSUES AUDIT');
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log('FOUND:', orphanIssues.length, 'orphan issues (exist in GitHub, missing in ROADMAP)\n');

  orphanIssues.forEach(i => {
    console.log(`#${i.number} - ${i.title.substring(0, 70)}`);
    console.log(`├─ State: ${i.state.toLowerCase()}`);
    console.log(`├─ Milestone: ${i.milestone?.title || 'None'}`);
    console.log(`└─ Action: ${i.state === 'CLOSED' ? 'Add to ROADMAP and mark as completed' : 'Add to ROADMAP pending section'}\n`);
  });
}

// Milestone validation
console.log('\n📈 MILESTONE PROGRESS AUDIT');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

const milestones = {
  'M1: Foundation - Testes': { roadmapProgress: '34/34 (100%)', line: 15 },
  'M2: CI/CD Pipeline': { roadmapProgress: '10/10 (100%)', line: 16 },
  'M3: Quality & Security': { roadmapProgress: '9/13 (69%)', line: 17 },
  'M4: Refactoring & Performance': { roadmapProgress: '4/20 (20%)', line: 18 },
  'M5: E2E Testing & Documentation': { roadmapProgress: '1/17 (6%)', line: 19 },
  'M6: Maintenance (Recurring)': { roadmapProgress: '0/2 (0%)', line: 20 }
};

console.log('| Milestone | ROADMAP    | GitHub     | Sync |');
console.log('|-----------|------------|------------|------|');

Object.entries(milestones).forEach(([name, info]) => {
  const issues = githubIssues.filter(i => i.milestone?.title === name);
  const total = issues.length;
  const closed = issues.filter(i => i.state === 'CLOSED').length;
  const pct = total > 0 ? Math.round((closed / total) * 100) : 0;
  const githubProgress = `${closed}/${total} ${pct}%`;
  const sync = info.roadmapProgress.includes(githubProgress.replace(/ /g, '')) ? '✅' : '❌';
  console.log(`| ${name.padEnd(30)} | ${info.roadmapProgress.padEnd(10)} | ${githubProgress.padEnd(10)} | ${sync.padEnd(4)} |`);
});

// State mismatches - check M3 issues mentioned in ROADMAP as closed
console.log('\n🔍 ISSUE STATE AUDIT (Sample)');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

// Check specific issues mentioned as completed in ROADMAP
const roadmapClosedPattern = /✅\s*#(\d+)/g;
const roadmapClosedMatches = roadmapContent.matchAll(roadmapClosedPattern);
const roadmapClosedIssues = new Set();
for (const match of roadmapClosedMatches) {
  roadmapClosedIssues.add(parseInt(match[1]));
}

let mismatches = 0;
[...roadmapClosedIssues].sort((a,b) => a-b).forEach(num => {
  const issue = githubIssues.find(i => i.number === num);
  if (issue && issue.state === 'OPEN') {
    console.log(`❌ #${num} - Marked ✅ in ROADMAP, but OPEN in GitHub`);
    console.log(`   └─ Title: ${issue.title.substring(0, 60)}`);
    console.log(`   └─ Milestone: ${issue.milestone?.title || 'None'}`);
    console.log(`   └─ Action: Either close in GitHub OR remove ✅ from ROADMAP\n`);
    mismatches++;
  }
});

if (mismatches === 0) {
  console.log('✅ All issues marked as completed (✅) in ROADMAP are actually CLOSED in GitHub');
}

// Velocity
console.log('\n⏱️  VELOCITY & ETA AUDIT');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n');

const now = new Date();
const sevenDaysAgo = new Date(now - 7 * 24 * 60 * 60 * 1000);
const recentlyClosed = githubIssues.filter(i => {
  if (i.state !== 'CLOSED' || !i.closedAt) return false;
  return new Date(i.closedAt) > sevenDaysAgo;
});

console.log('ACTUAL VELOCITY (Last 7 days):');
console.log('├─ Issues closed:', recentlyClosed.length, 'issues');
console.log('├─ Average:', (recentlyClosed.length / 7).toFixed(1), 'issues/day');
console.log('└─ Trend: Velocity data from last 7 days');

const remainingOpen = githubIssues.filter(i => i.state === 'OPEN').length;
const velocity = recentlyClosed.length / 7;
const daysToComplete = velocity > 0 ? (remainingOpen / velocity).toFixed(1) : 'N/A';
console.log('\nPROJECTED COMPLETION:');
console.log('├─ Remaining issues:', remainingOpen);
console.log('├─ Current velocity:', velocity.toFixed(1), 'issues/day');
console.log('└─ Est. days to complete:', daysToComplete, 'days');

if (daysToComplete !== 'N/A') {
  const completionDate = new Date(now.getTime() + parseFloat(daysToComplete) * 24 * 60 * 60 * 1000);
  console.log('    ETA:', completionDate.toISOString().split('T')[0]);
}

// Summary
console.log('\n═══════════════════════════════════════════════════════');
console.log('✅ AUDIT COMPLETE');
console.log('═══════════════════════════════════════════════════════');

File: roadmap-audit.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
ROADMAP AUDIT SCRIPT
Cross-reference GitHub issues with ROADMAP.md
"""

import json
import re
from datetime import datetime, timedelta
from collections import defaultdict

# Load GitHub issues
with open('github-issues.json', 'r', encoding='utf-8') as f:
    issues = json.load(f)

# Load ROADMAP
with open('ROADMAP.md', 'r', encoding='utf-8') as f:
    roadmap = f.read()

print("═" * 63)
print("🎯 ROADMAP AUDIT - COMPREHENSIVE ANALYSIS")
print("═" * 63)
print()
print(f"Audit Date: {datetime.now().strftime('%Y-%m-%d')}")
print(f"Scope: {len(issues)} GitHub issues vs ROADMAP.md")
print()

# ============================================================================
# SECTION 1: ISSUE COUNT RECONCILIATION
# ============================================================================
print("📊 SECTION 1: ISSUE COUNT RECONCILIATION")
print("━" * 63)
print()

github_total = len(issues)
github_open = sum(1 for i in issues if i['state'] == 'OPEN')
github_closed = sum(1 for i in issues if i['state'] == 'CLOSED')

# Extract ROADMAP claimed total
roadmap_match = re.search(r'Progresso Global.*?(\d+)/(\d+)', roadmap)
if roadmap_match:
    roadmap_closed = int(roadmap_match.group(1))
    roadmap_total = int(roadmap_match.group(2))
else:
    roadmap_closed = 0
    roadmap_total = 0

drift = github_total - roadmap_total
drift_pct = round((drift / roadmap_total) * 100, 1) if roadmap_total > 0 else 0

print(f"ROADMAP.md:        {roadmap_total} issues ({roadmap_closed} closed)")
print(f"GitHub (actual):   {github_total} issues ({github_closed} closed)")
print(f"Drift:             {'+' if drift > 0 else ''}{drift} issues ({drift_pct}%)")

status = "🔴 CRITICAL" if abs(drift_pct) > 5 else "🟡 WARNING" if abs(drift_pct) > 2 else "🟢 ACCEPTABLE"
print(f"Status:            {status} drift")
print()

print("BREAKDOWN:")
print(f"  ✅ GitHub Total:          {github_total} issues")
print(f"  📝 ROADMAP Documented:    {roadmap_total} issues")
print(f"  ❌ Discrepancy:           {'+' if drift > 0 else ''}{drift} issues")
print()

# ============================================================================
# SECTION 2: MILESTONE PROGRESS VALIDATION
# ============================================================================
print("📈 SECTION 2: MILESTONE PROGRESS VALIDATION")
print("━" * 63)
print()

milestones_data = [
    {"name": "M1", "title": "M1: Foundation - Testes", "roadmap_total": 35, "roadmap_closed": 35},
    {"name": "M2", "title": "M2: CI/CD Pipeline", "roadmap_total": 18, "roadmap_closed": 18},
    {"name": "M3", "title": "M3: Quality & Security", "roadmap_total": 57, "roadmap_closed": 57},
    {"name": "M4", "title": "M4: Refactoring & Performance", "roadmap_total": 44, "roadmap_closed": 44},
    {"name": "M5", "title": "M5: E2E Testing & Documentation", "roadmap_total": 22, "roadmap_closed": 2},
    {"name": "M6", "title": "M6: Maintenance (Recurring)", "roadmap_total": 11, "roadmap_closed": 2},
    {"name": "M7", "title": "M7: Multi-Tenancy B2G", "roadmap_total": 6, "roadmap_closed": 0}
]

print("| Milestone | ROADMAP        | GitHub         | Sync | Notes                   |")
print("|-----------|----------------|----------------|------|-------------------------|")

for ms in milestones_data:
    gh_issues = [i for i in issues if i.get('milestone') and i['milestone']['title'] == ms['title']]
    gh_total = len(gh_issues)
    gh_closed = sum(1 for i in gh_issues if i['state'] == 'CLOSED')
    gh_pct = round((gh_closed / gh_total) * 100) if gh_total > 0 else 0

    roadmap_pct = round((ms['roadmap_closed'] / ms['roadmap_total']) * 100) if ms['roadmap_total'] > 0 else 0

    if gh_total == ms['roadmap_total'] and gh_closed == ms['roadmap_closed']:
        sync = "✅"
    elif abs(gh_closed - ms['roadmap_closed']) <= 2:
        sync = "⚠️"
    else:
        sync = "❌"

    roadmap_str = f"{ms['roadmap_closed']}/{ms['roadmap_total']} {roadmap_pct}%"
    gh_str = f"{gh_closed}/{gh_total} {gh_pct}%"

    if gh_total != ms['roadmap_total']:
        note = f"Count mismatch: {gh_total - ms['roadmap_total']:+d}"
    elif gh_closed != ms['roadmap_closed']:
        note = "Progress mismatch"
    else:
        note = "Perfect sync"

    print(f"| {ms['name']:<9} | {roadmap_str:<14} | {gh_str:<14} | {sync}    | {note:<23} |")

print()

# ============================================================================
# SECTION 3: ISSUE NUMBER ANALYSIS
# ============================================================================
print("🔍 SECTION 3: ISSUE NUMBER ANALYSIS")
print("━" * 63)
print()

all_issue_numbers = sorted([i['number'] for i in issues])
print(f"GitHub Issue Range: #{all_issue_numbers[0]} to #{all_issue_numbers[-1]}")
print(f"Total Issues in GitHub: {len(all_issue_numbers)}")
print()

# Extract all issue references from ROADMAP
issue_pattern = r'#(\d+)'
matches = re.findall(issue_pattern, roadmap)
documented_issues = sorted(list(set([int(m) for m in matches if int(m) <= 400])))

print(f"Issues Referenced in ROADMAP: {len(documented_issues)}")
print()

# Find orphans (in GitHub but not in ROADMAP)
orphans = [n for n in all_issue_numbers if n not in documented_issues]

print(f"🆕 ORPHAN ISSUES (In GitHub, NOT in ROADMAP): {len(orphans)}")
if orphans:
    orphan_details = [i for i in issues if i['number'] in orphans]
    orphan_details.sort(key=lambda x: x['number'])

    for orphan in orphan_details[:20]:
        ms = orphan.get('milestone', {}).get('title', 'No Milestone') if orphan.get('milestone') else 'No Milestone'
        print(f"  #{orphan['number']} - [{orphan['state'].upper()}] {orphan['title']}")
        print(f"    Milestone: {ms}")

    if len(orphans) > 20:
        print(f"  ... and {len(orphans) - 20} more orphan issues")
print()

# Find phantoms (in ROADMAP but not in GitHub)
phantoms = [n for n in documented_issues if n not in all_issue_numbers]

print(f"👻 PHANTOM ISSUES (In ROADMAP, NOT in GitHub): {len(phantoms)}")
if phantoms:
    print(f"  Phantom issue numbers: {', '.join(f'#{n}' for n in phantoms)}")
print()

# ============================================================================
# SECTION 4: VELOCITY & ETA
# ============================================================================
print("⏱️ SECTION 4: VELOCITY & ETA VALIDATION")
print("━" * 63)
print()

from datetime import timezone
seven_days_ago = datetime.now(timezone.utc) - timedelta(days=7)
closed_last_7 = [i for i in issues if i.get('closedAt') and datetime.fromisoformat(i['closedAt'].replace('Z', '+00:00')) > seven_days_ago]
velocity = round(len(closed_last_7) / 7, 1)

print("ACTUAL VELOCITY (Last 7 days):")
print(f"  Issues closed: {len(closed_last_7)} issues")
print(f"  Average: {velocity} issues/day")
print()

# Extract ROADMAP claimed velocity
velocity_match = re.search(r'Velocidade:\*\*\s*(\d+\.?\d*)\s*issues/dia', roadmap)
if velocity_match:
    roadmap_velocity = float(velocity_match.group(1))
    print(f"ROADMAP Claimed Velocity: {roadmap_velocity} issues/day")
    velocity_diff = velocity - roadmap_velocity
    print(f"Velocity Accuracy: {'+' if velocity_diff > 0 else ''}{velocity_diff} issues/day")
print()

# Remaining work
remaining_open = github_open
days_to_complete = round(remaining_open / velocity, 1) if velocity > 0 else 0
projected_end = (datetime.now() + timedelta(days=days_to_complete)).strftime('%Y-%m-%d')

print("PROJECTION:")
print(f"  Remaining open issues: {remaining_open}")
print(f"  Days to complete: {days_to_complete} days (at current velocity)")
print(f"  Projected completion: {projected_end}")
print()

# ============================================================================
# SECTION 5: SUMMARY
# ============================================================================
print("═" * 63)
print("✅ AUDIT SUMMARY")
print("═" * 63)
print()

print("KEY FINDINGS:")
print(f"  1. Issue Count Drift: {'+' if drift > 0 else ''}{drift} issues ({drift_pct}%)")
print(f"  2. Orphan Issues: {len(orphans)} issues in GitHub not documented")
print(f"  3. Phantom Issues: {len(phantoms)} issues documented but don't exist")
print(f"  4. Velocity: {velocity} issues/day (last 7 days)")
print(f"  5. Projected Completion: {projected_end}")
print()

print("SYNC STATUS: ", end='')
if abs(drift_pct) < 2 and len(orphans) == 0 and len(phantoms) == 0:
    print("🟢 EXCELLENT (< 2% drift, no orphans/phantoms)")
elif abs(drift_pct) < 5 and len(orphans) <= 5 and len(phantoms) == 0:
    print("🟡 ACCEPTABLE (< 5% drift, minor orphans)")
else:
    print("🔴 NEEDS ATTENTION (>5% drift or issues detected)")
print()

print("═" * 63)
print("AUDIT COMPLETE")
print("═" * 63)
